### PR TITLE
EOS-9010: File handle support for cortxfs API's (cortfx-posix repo)

### PR DESCRIPTION
**Pull request by shahidSTX**
# cortxfs-posix Change Summary

## Problem Statement
_[EOS-9010](https://jts.seagate.com/browse/EOS-9010):_
_File handle support for cortxfs API's_

## Problem Description
_The current implementation relies on cfs_fs and inode, that should be eliminated_

## Solution Overview
_Modified relevant API's to use FH_

## Unit Test Cases
_Build, UT testing, ensure no errors_

## Testing
<details>
  <summary>Click here to see the build output</summary>

```
[root@ssc-vm-c-0689 cortx-posix]# ./scripts/test.sh
 --
 --
Configuration Completed
Clean indexes prepared
NSAL Unit tests
NS Tests
Test results are logged to /var/log/cortx/test/ut/ut_nsal.log
Total tests  = 5
Tests passed = 5
Tests failed = 0

Iterator test
Test results are logged to /var/log/cortx/test/ut/ut_nsal.log
Total tests  = 2
Tests passed = 2
Tests failed = 0

KVTree test
Test results are logged to /var/log/cortx/test/ut/ut_nsal.log
Total tests  = 17
Tests passed = 17
Tests failed = 0

Global KVS Tests
Test results are logged to /var/log/cortx/test/ut/ut_nsal.log
Total tests  = 4
Tests passed = 4
Tests failed = 0

CORTXFS Unit tests
Endpoint ops Tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.logs
Total tests  = 4
Tests passed = 4
Tests failed = 0

FS Tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 3
Tests passed = 3
Tests failed = 0

Directory tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 14
Tests passed = 14
Tests failed = 0

File creation tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 4
Tests passed = 4
Tests failed = 0

Link Tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 6
Tests passed = 6
Tests failed = 0

Rename tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 3
Tests passed = 3
Tests failed = 0

Attribute Tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 5
Tests passed = 5
Tests failed = 0

Xattr file Tests
Test results are logged to /var/log/cortx/test/ut/xattr_file_ops.log
Total tests  = 9
Tests passed = 9
Tests failed = 0

Xattr dir Tests
Test results are logged to /var/log/cortx/test/ut/xattr_dir_ops.log
Total tests  = 9
Tests passed = 9
Tests failed = 0

IO tests
Test results are logged to /var/log/cortx/test/ut/ut_cortxfs.log
Total tests  = 7
Tests passed = 7
Tests failed = 0

DSAL Unit tests
Dsal basic test
Test results are logged to /var/log/cortx/test/ut/ut_dsal.log
Total tests  = 11
Tests passed = 11
Tests failed = 0

Dsal IO test
Test results are logged to /var/log/cortx/test/ut/ut_dsal.log
Total tests  = 2
Tests passed = 2
Tests failed = 0

Dsal space stats test
Test results are logged to /var/log/cortx/test/ut/ut_dsal.log
Total tests  = 2
Tests passed = 2
Tests failed = 0

```

</details>

## Checklist
- [x] **Compilation:** _This patch does not break compilation_
- [x] **Merge conflicts:** _None_
- [ ] **Code review:** _In progress_
- [x] **Sanity Testing:** _Completed as mentioned above_
- [ ] **Documentation:** _TBD_

Relevant PR's for this :
https://github.com/Seagate/cortx-fs/pull/70
https://github.com/Seagate/cortx-fs-ganesha/pull/59

### CODE REVIEW

File: experiments/getattr_profiling.c [hunk 1]
```diff
@@ -85,13 +88,12 @@ static void set_ctime(void **state)
 	gettimeofday(&st,NULL);
 	for (i=0;i<NUM_FILES;i++)
 	{
-		memset(&stat_out, 0, sizeof(stat_out));
 		ut_cfs_objs->file_inode=ut_dir_obj->file_inode[i];
-		rc = cfs_getattr(ut_cfs_objs->cfs_fs, &ut_cfs_objs->cred,
-				&ut_cfs_objs->file_inode, &stat_out);
+		rc = cfs_fh_from_ino(ut_cfs_obj->cfs_fs,
```
Comment (Anilsindhi) [inline experiments/getattr_profiling.c:92]: Here you are over writing FH with newly constructed FH and hence loosing the track of old FH.
1. If you created a FH locally you have to free it at the end once you are done, so you can take array of FH if you have multiple FH to be constructed, it is a memory leak here. And destruct them at the end when not needed. 
2. Seems like you are creating the same FH in two different for loop, but you could have preserved the FHs created in first for loop and could have used those later on.
3. I doubt how this logic is still working, as you created an in memory FH, in that you updated stat. In second step(Second for loop) you are again getting FH from backend(We did not pushed the FH content back to backend in first for loop) so you should be getting old stat not updated and hence your assert should have failed here. If we do the changes suggested in point 2 above, it should not happen.
3. Looks like same way it is done at multiple place, so needs to be corrected at other place too.

**Review by Anilsindhi — COMMENTED** [review]
Few minor comments, please have a look at them.

File: experiments/getattr_profiling.c (hunk 1 above)
Comment (shahidSTX) [reply experiments/getattr_profiling.c:92]: Point 1 and 2 seem to be valid, I will address them by using array of FH's, re-using them in second for loop and then freeing them up at last.
Point 3 : this code might not be exercised as of now, hence we don't know if any of this logic is working.
Point 4 : will address.

File: experiments/getattr_profiling.c [hunk 2]
```diff
@@ -218,8 +222,10 @@ static void set_atime(void **state)
 		time(&cur_time[i]);
 
 		ut_cfs_objs->file_inode=ut_dir_obj->file_inode[i];
-		rc = cfs_setattr(ut_cfs_objs->cfs_fs, &ut_cfs_objs->cred,
-				 &ut_cfs_objs->file_inode, &stat_in, flag);
+		rc = cfs_fh_from_ino(ut_cfs_obj->cfs_fs,
+					&ut_cfs_objs->file_inode, &fh[i]);
```
Comment (Anilsindhi) [inline experiments/getattr_profiling.c:226]: Still not addressed the alignment.

File: experiments/getattr_profiling.c [hunk 3]
```diff
@@ -284,9 +291,10 @@ static void set_gid(void **state)
 		time(&cur_time);
 
 		ut_cfs_objs->file_inode=ut_dir_obj->file_inode[i];
-		rc = cfs_setattr(ut_cfs_objs->cfs_fs, &ut_cfs_objs->cred,
-				&ut_cfs_objs->file_inode, &stat_in, flag);
-
+		rc = cfs_fh_from_ino(ut_cfs_obj->cfs_fs,
+					&ut_cfs_objs->file_inode, &fh[i]);
```
Comment (Anilsindhi) [inline experiments/getattr_profiling.c:295]: Align it properly.

File: experiments/getattr_profiling.c [hunk 4]
```diff
@@ -350,8 +357,10 @@ static void set_uid(void **state)
 		time(&cur_time[i]);
 
 		ut_cfs_objs->file_inode=ut_dir_obj->file_inode[i];
-		rc = cfs_setattr(ut_cfs_objs->cfs_fs, &ut_cfs_objs->cred,
-				&ut_cfs_objs->file_inode, &stat_in, flag);
+		rc = cfs_fh_from_ino(ut_cfs_obj->cfs_fs,
+					&ut_cfs_objs->file_inode, &fh[i]);
```
Comment (Anilsindhi) [inline experiments/getattr_profiling.c:361]: Align it properly.

File: experiments/getattr_profiling.c [hunk 5]
```diff
@@ -74,8 +75,10 @@ static void set_ctime(void **state)
 
 		stat_in.st_ctim.tv_sec = new_ctime;
 		ut_cfs_objs->file_inode=ut_dir_obj->file_inode[i];
-		rc = cfs_setattr(ut_cfs_objs->cfs_fs, &ut_cfs_objs->cred,
-				&ut_cfs_objs->file_inode, &stat_in, flag);
+		rc = cfs_fh_from_ino(ut_cfs_obj->cfs_fs,
+					&ut_cfs_objs->file_inode, &fh);
```
[the remark this answers is not in the document]
Comment (Anilsindhi) [reply experiments/getattr_profiling.c]: This line is still not aligned it should have one more space

File: experiments/getattr_profiling.c (hunk 1 above)
Comment (Anilsindhi) [reply experiments/getattr_profiling.c]: Thanks for addressing this.

File: experiments/getattr_profiling.c [hunk 6]
```diff
@@ -145,9 +148,10 @@ static void set_mtime(void **state)
 
 		time(&cur_time[i]);
 		ut_cfs_objs->file_inode=ut_dir_obj->file_inode[i];
-		rc = cfs_setattr(ut_cfs_objs->cfs_fs, &ut_cfs_objs->cred,
-				 &ut_cfs_objs->file_inode, &stat_in, flag);
-
+		rc = cfs_fh_from_ino(ut_cfs_obj->cfs_fs,
+				    &ut_cfs_objs->file_inode, &fh[i]);
```
Comment (Anilsindhi) [inline experiments/getattr_profiling.c:152]: Align this properly. Just  one more space.

**Review by Anilsindhi — COMMENTED** [review]
1. Please post the output of getattr_profiling unit test in PR description.
2. Address the minor comment given for alignment.

File: experiments/getattr_profiling.c (hunk 6 above)
Comment (shahidSTX) [reply experiments/getattr_profiling.c:152]: Done

File: experiments/getattr_profiling.c (hunk 2 above)
Comment (shahidSTX) [reply experiments/getattr_profiling.c:226]: Done.

File: experiments/getattr_profiling.c (hunk 3 above)
Comment (shahidSTX) [reply experiments/getattr_profiling.c:295]: Done

File: experiments/getattr_profiling.c (hunk 4 above)
Comment (shahidSTX) [reply experiments/getattr_profiling.c:361]: Done.

**Comment by shahidSTX** [comment]
This branch is discarded and new changes are here :

https://github.com/Seagate/cortx-posix/pull/299
https://github.com/Seagate/cortx-fs/pull/74
https://github.com/Seagate/cortx-fs-ganesha/pull/60

**Pull request closed without merging**